### ISSUE TITLE
[$250] Safari mWeb - Keyboard won't open for some inputs (New Chat, Change Password, New Room)

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#10414]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/10194

### case 1
1. Access staging.new.expensify.com
2. Sign into a valid user
3. Click on the "Global action" button (Big green plus button in LHN)
3. Click New Chat

### case 2
1. Go to https://staging.new.expensify.com/
2. Login with expensifail account
3. Create a Workspace
4. Tap "+"
5. Tap "New Room"

### case 3
1. Go to https://staging.new.expensify.com/
2. Sign into a valid user
3. Go to settings > security > change password

### case 4
2. Go to https://staging.new.expensify.com/
3. Sign into a valid user
4. Go to some DM
5. Open the in app emoji picker
6. Select some emoji

## Expected Result:
Keyboard is open and cursor focused when entering the new screen

## Actual Result:
 Keyboard won't open

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Safari Mobile Web

**Version Number:** 1.1.88.13

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

### Case 1
https://user-images.githubusercontent.com/93399543/185005745-73a26290-2721-4b14-b3bd-ba0553a2a0a8.mp4

### Case 2
https://user-images.githubusercontent.com/93399543/195383970-9354b2d0-cffc-4b1d-b5d2-2804c2462a5e.mp4

### Case 3
https://user-images.githubusercontent.com/43995119/190952432-d072befe-c01d-404e-92fd-b6c71ff230c9.mov

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Tea

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by NicMendonca** [comment]
job post: https://www.upwork.com/jobs/~01f21a08ef50dd0283

**Comment by busyboxww** [comment]
The text is not having top "priority" or the list is getting auto focus. You should give change the layout, something zIndex or ... set focus programmatically.

**Comment by NicMendonca** [comment]
@rushatgabhane are you recommending that I close this?

**Comment by kbecciv** [comment]
@rushatgabhane Issue is reproduced with latest build 1.1.94.4

https://user-images.githubusercontent.com/93399543/187564523-ea61b4d2-b8ff-4d5f-a70f-2d42605dc211.MP4

**Comment by NicMendonca** [comment]
Still reproducible ☝️

**Comment by sketchydroide** [comment]
@parasharrajat did you have time to look into this?

**Comment by sketchydroide** [comment]
no sorry, it's weird, but I' sure I did not have you comment with the screenshot when I wrote my comment, maybe GH did not refresh it, weird.

**Comment by NicMendonca** [comment]
price doubled - https://www.upwork.com/jobs/~01f21a08ef50dd0283

**Comment by NicMendonca** [comment]
doubled: https://www.upwork.com/jobs/~01f21a08ef50dd0283

**Comment by mallenexpensify** [comment]
Is the issue I'm facing the same as this one? 
On desktop, staging, v1.2.15-2
- Follow repro steps in OP
- After clicking a new chat, cursor isn't focused on compose box, unable to start typing immediately, have to click compose box to do so. 

Same for when I click on a chat in LHN, cursor doesn't show in compose box

**Comment by parasharrajat** [comment]
That sounds like a new issue and I saw a bug report for it somewhere. This issue is concerned on the search box on the new chat and new group page.

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat https://expensify.slack.com/archives/C01GTK53T8Q/p1665792005925189

**Comment by NicMendonca** [comment]
doubled price: https://www.upwork.com/jobs/~01f21a08ef50dd0283

**Comment by NicMendonca** [comment]
@parasharrajat maybe open up a discussion in slack?

**Comment by parasharrajat** [comment]
@sketchydroide What do you think of https://github.com/Expensify/App/issues/10414#issuecomment-1285514867?

**Comment by sketchydroide** [comment]
They do feel like all being the same issue @parasharrajat I think we can merge them

**Comment by parasharrajat** [comment]
Who can help with that? I don't have access to do that.

**Comment by sketchydroide** [comment]
I think I can do it
It's just closing the other and mantainng one, also probably updating the description, let me see if I can do that myself.

**Comment by puneetlath** [comment]
@NicMendonca @sketchydroide it sounds like we're closing this issue, is that right? If so, can one of you go ahead and do it?

**Comment by parasharrajat** [comment]
We are not closing this one. Instead, merge other issues into this and close those. I don't have access to modify it to I am waiting on someone to take up on that.

**Comment by sketchydroide** [comment]
yep the others are now closed

**Comment by sketchydroide** [comment]
lets continue this here

**Comment by parasharrajat** [comment]
Thanks, @sketchydroide. Could you please update the description to add more details about other issues? Adding those pages and steps to those issues will be enough. 

We can also generalize the Expected behavior and title.

**Comment by JmillsExpensify** [comment]
@tgolen Another one to put on our radar. Shall we add it to the tracking issue?

**Comment by sketchydroide** [comment]
>Thanks, @sketchydroide. Could you please update the description to add more details about other issues? Adding those pages and steps to those issues will be enough.


Sure @parasharrajat

**Comment by sketchydroide** [comment]
Ok Updated, I'm not sure if the fact that we have 3 instances where this happens can be a bit confusing on the description, but I'll take any suggestions on how we can improve that.

**Comment by tgolen** [comment]
> @tgolen Another one to put on our radar. Shall we add it to the tracking issue?

It's probably a different problem than the ones in that tracking issue. I think this is more of an auto-focus problem, rather than a problem with views being obscured by the keyboard.

**Comment by JmillsExpensify** [comment]
Gotcha, thanks for the second opinion!

**Comment by NicMendonca** [comment]
Updated the job name on Upwork, and doubled the price: https://www.upwork.com/jobs/~01f21a08ef50dd0283

**Comment by kavimuru** [comment]
#12439

**Comment by NicMendonca** [comment]
I am about to go OOO for a week, so re-assigning.

**Comment by s77rt** [comment]
@kbecciv @parasharrajat Can you still reproduce? I can only reproduce the "New Chat" issue but I think it's related to another root cause

**Comment by sketchydroide** [comment]
waiting for some feedback, it would be weird if this just fixed by itself, but maybe a react update could do it?

**Comment by kbecciv** [comment]
@sketchydroide @s77rt Checking, will update you shortly

**Comment by parasharrajat** [comment]
It will take some time to review a few pages. This must be affecting more pages than listed on the issue. 

Funny thing: All of these were working fine a few months back. (Needs RCA). 

Stay tuned.

**Comment by s77rt** [comment]
@parasharrajat 
> Funny thing: All of these were working fine a few months back. (Needs RCA).
I think you are wrong. Nothing was working. And you just reminded me, the reproduced issue you showed [here](https://github.com/Expensify/App/issues/10414#issuecomment-1238555503) is not related to this issue. In your case the input is not focused at all. On our cases the input is focused (kinda) but the keyboard does not open. My proposal fixes your issue too.
I see you are using a simulator, to reproduce the issue on the simulator you have to disable hardware keyboard.
I have already provided the RCA and the "Rules" to make it work.

**Comment by s77rt** [comment]
@kbecciv sorry for the ping, can you apply my proposal and retest?

**Comment by kbecciv** [comment]
Sorry, @s77rt This is not my job.

**Comment by bulatgab** [comment]
@s77rt FYI, it fixed the issue on my end.
(It doesn't matter much though - I'm not an Expensify employee, I'm just a fellow contributor and I was also looking into this issue. You will need to wait for someone else to answer 🙂)

**Comment by s77rt** [comment]
@bulatgab Thanks for taking the time and testing. Was it on real hardware or simulation?

**Comment by bulatgab** [comment]
It was on a simulator (iPhone 13, iOS 16.1). I applied your changes to the latest staging `3bf0b729`

**Comment by parasharrajat** [comment]
I will apply the changes tomorrow to test them.

**Comment by parasharrajat** [comment]
@michaelhaxhiu Do we have to hold this for https://expensify.slack.com/archives/C01GTK53T8Q/p1668604822482759?

**Comment by michaelhaxhiu** [comment]
Let's have the assigned BZ help drive this forward (@kevinksullivan). We likely need to make a judgement call here about proceeding forward with the proposal or closing/paying a % for the involved.

**Comment by JmillsExpensify** [comment]
@sketchydroide To help clear out the extensive backlog of /App bugs, we're putting the spotlight on all bugs older than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:

 - Decide whether any proposals currently meet our guidelines and can be approved as-is
 - For any that can't, please take this issue internal and treat it as one of your highest priorities
 - If you have any questions, don't hesitate to start a discussion in #bug-zero

Thanks everyone!

**Comment by JmillsExpensify** [comment]
@sketchydroide In addition to that message, clarifying that for your specific case please focus on IOU refactors and then this issue.

**Comment by parasharrajat** [comment]
@kevinksullivan Are we holding this one as per https://expensify.slack.com/archives/C01GTK53T8Q/p1668604822482759?

**Comment by pecanoro** [comment]
I tried to reproduce this on mobile web and the native app on Android to see if I could help with a fix but it works just fine, I think we should close it.

**Comment by s77rt** [comment]
@pecanoro 
> First of all, it's important to understand that the issue is only related to Mobile Safari.

https://github.com/Expensify/App/issues/10414#issuecomment-1314058212

**Comment by pecanoro** [comment]
Then let's update the description to say mobile safari only, because it's pretty confusing otherwise.

**Comment by parasharrajat** [comment]
Bump @kevinksullivan  https://github.com/Expensify/App/issues/10414#issuecomment-1322612339

**Comment by tjferriss** [comment]
I'm doing the weekly bug review and based on the linked Slack thread this looks like a hold so I updated the title. @kevinksullivan if you disagree then let's chat in the Slack thread.

**Comment by tjferriss** [comment]
Whoops sorry, looks like we shouldn't be putting this on hold based on this [Slack discussion](https://expensify.slack.com/archives/C01GTK53T8Q/p1669139081715229).

**Comment by sketchydroide** [comment]
I haven't been able to look at this today, will definetly do tomorrow

**Comment by parasharrajat** [comment]
I am not able to reproduce this. Installing SDK 16 to test.

**Comment by s77rt** [comment]
@parasharrajat 
> I see you are using a simulator, to reproduce the issue on the simulator you have to disable hardware keyboard.

**Comment by parasharrajat** [comment]
I can reproduce it now. Took some time to cleanup the old Xcode and install the new one.

**Comment by sketchydroide** [comment]
@parasharrajat any updates on testing @s77rt suggestion?

**Comment by parasharrajat** [comment]
Yeah, I do. Trying to test it first but it will take some time my mac is broken again, Thanks to me.

**Comment by sketchydroide** [comment]
> my mac is broken again, Thanks to me.

😭

**Comment by parasharrajat** [comment]
There are some bootup issues so it shouldn't take much.

**Comment by s77rt** [comment]
@parasharrajat
The change to the modal is just to not keep the modal covering the entire screen otherwise we won't be able to focus on other elements that are under the modal.
Check rule number 3 on https://github.com/Expensify/App/issues/10414#issuecomment-1314058212

**Comment by JmillsExpensify** [comment]
@sketchydroide @parasharrajat We really need to speed up evaluating @s77rt's proposal. It was provided 15 days ago. Can we close the loop on this today?

**Comment by sketchydroide** [comment]
Yes, sorry his proposal got lost in the midst of the colapsable items on GH, we will try to get this reviewed asap

**Comment by kevinksullivan** [comment]
Just to clarify @sketchydroide you will get this reviewed today (your time?)

**Comment by sketchydroide** [comment]
I will try to organise myself with @parasharrajat to review the proposal, my main priority for the day

**Comment by parasharrajat** [comment]
I have fixed the Mac. So it will be reviewed today. Currently reviewing a PR and I will take it up next.

**Comment by parasharrajat** [comment]
> https://github.com/Expensify/App/issues/10414#issuecomment-1330912143
 
@s77rt Can you please share the video of the issue that it is solving?

**Comment by s77rt** [comment]
@parasharrajat 
https://github.com/Expensify/App/issues/10414#issuecomment-1314058212
New Chat, New Room, New Group, and basically any other pages that are initialled from the modal (after clicking the green + circle)

**Comment by sketchydroide** [comment]
> > [#10414 (comment)](https://github.com/Expensify/App/issues/10414#issuecomment-1330912143)
> 
> @s77rt Can you please share the video of the issue that it is solving?

Not sure what you mean by this @parasharrajat there is a video on the comment, do you need something else/more?

>New Chat, New Room, New Group, and basically any other pages that are initialled from the modal (after clicking the green + circle)

@s77rt also not sure how this answers @parasharrajat, are you asking which pages do you need to do a video from?

**Comment by sketchydroide** [comment]
I've also checked the diff proposed by @s77rt and it looks good as far as I can tell. I think once we test it and have proof that this is working we can assign it and start the PR

**Comment by s77rt** [comment]
@sketchydroide I think he asked why i made changes to the modal
@parasharrajat I have summarised the 3 rules to follow and for each situation I listed which rules are not being followed and proposed a fix for each rule

**Comment by parasharrajat** [comment]
Yeah, let me run it first. I am assuming rule 3 is for web not mWeb.

**Comment by s77rt** [comment]
@parasharrajat 
> Yeah, let me run it first. I am assuming rule 3 is for web not mWeb.

No; the rules are for both web and mWeb, there is no diff between the two. It just happens that Safari behaves differently on Web (since there is no virtual keyboard, it follows a bit different principles I think)

**Comment by s77rt** [comment]
Indeed rule 3 is about trap focus, I had found that while working on a similar case https://github.com/Expensify/App/issues/12521 where I had to re-implement a [simple trap focus](https://github.com/Expensify/App/issues/12521#issuecomment-1319343035) instead of using RN's but that was due to another reason which may not apply here.

The rest of the rules are still true, those are rules and not in-depth technical analysis, for instance rule 1 and rule 2 are the same, just had to split them for clarity

**Comment by marcaaron** [comment]
Can someone please summarize in as few words as possible what we need to do here? I am looking at the recent posts from @s77rt and @ntdiary and honestly it's just super hard for me to follow either train of thought. I would suggest we stop using the "rules" terminology and just refer to whatever it is we are talking about instead of "rule 1", "rule 2", etc.

**Questions...**

- I'm getting that the bug is related to the backdrop on a modal somehow. But which modal are we talking about? 

- Does the presence of a "backdrop" prevent the focus from happening?  How?

- What are `trapFocus`, `coverScreen` and `hasBackdrop` all used for?

- If iOS Safari only opens the keyboard in response to a "user interaction" then why does the keyboard open at all?

Also, I think it would be helpful to stop sharing code proposals until the root problem is understood better.

**Comment by s77rt** [comment]
@marcaaron In short, we need to focus on the inputs right on user clicking actions without callbacks, timeouts, etc. And that textinput should be focus-able and not behind a modal that implements trap focus

**Answers**

- No, it's not, it's related to `coverScreen`. We are referring to all modals generally but for this specific case we are referring to the modal that appears after clicking on the green circle (+) button


- It's about `coverScreen` Setting it to `true` will use RN's modal and it implements trap focus by defaults. Setting it to `false` will use react-native-modal implemented `<View />` and it does not implement trap focus


- `trapFocus` does not exist
`coverScreen` whether to use RN modal (append a div to the root document and cover the screen) or react-native-modal modal (append a div as a child of the parent element)
`hasBackdrop` just the backdrop (visual effect) nothing else


- The keyboard does not open

**Comment by marcaaron** [comment]
So following the thought that Safari on iOS is restrictive about when it allows calling `focus()` on a text input. If it _has to_ be tied to a user click interaction (I think this is what @ntdiary is saying and the link they provided). The part I'm struggling with is how we are supposed to `focus()` something that isn't on the screen yet. 

The first tap on something like "New Chat" will move us to that page. When it mounts the text input will _eventually_ be ready to get "focused" programmatically - but probably is not ready when the click handler for the FAB fires.

I am having doubts about whether this is something we can fix without a workaround. We could complain to Apple, but it's their choice whether to allow "async" focusing of inputs or not.

**Comment by s77rt** [comment]
We just must do it on the same call stack, it's okay if it takes long, just don't focus using settimeout or such...
Check my proposal above https://github.com/Expensify/App/issues/10414#issuecomment-1314058212
Using autoFocus works, as long as we render the text input at the firsr render call so we keep the same call stack

**Comment by marcaaron** [comment]
>we need to focus on the inputs right on user clicking actions without callbacks, timeouts, etc

How can we possibly do that if the input is not on screen when the user navigates somewhere?

>for this specific case we are referring to the modal that appears after clicking on the green circle (+) button

Sorry, which modal exactly? I don't see any modals blocking the input when it appears on the screen. Are you referring to the bottom docked menu?

>It's about coverScreen Setting it to true will use RN's modal and it implements trap focus by defaults. Setting it to false will use react-native-modal implemented <View /> and it does not implement trap focus

What does "it implements trap focus by defaults" mean? Not familiar with this - can you explain it another way?

**Comment by marcaaron** [comment]
It kind of sounds like we are proposing changes that will ensure that the call to `focus()` happens in the same tick of the event loop as the click event on the FAB menu. Am I understanding that part correctly?

**Comment by marcaaron** [comment]
@s77rt @ntdiary the solutions you provided reintroduce a performance related bug by removing the transition loader. There's a much bigger lag when navigating after and the page just pops into view instead of a smooth (well, honestly, it could be smoother) transition.

I think we could possibly solve this by moving the loader into the OptionsSelector itself and always rendering the input?

**Comment by marcaaron** [comment]
Hmm nope that doesn't work - it still looks bad with a big lag 😅

**Comment by marcaaron** [comment]
That being said... we are doing this on the search page and I think that's how it avoids the problem

https://github.com/Expensify/App/blob/820e876d5e4a8cbffe272220c1182c8942ed93cf/src/pages/SearchPage.js#L186

**Comment by marcaaron** [comment]
Tbh, I think we'd be ok with the solution to add `autoFocus` and use `shouldShowOptions` for the `NewChatPage` for consistency (as the `SearchPage` already has this). But I'm still struggling with the fact that this bug is related to some kind of "focus trap" from the popover menu in the default RN modal vs. the "implemented `View`" of `react-native-modal`. I still don't have a good idea of what is happening with that.

@parasharrajat you added these props in [this PR](https://github.com/Expensify/App/pull/3515/files) do you remember why? And is it safe to change them here? 

Coming back to this...

>It's about coverScreen Setting it to true will use RN's modal and it implements trap focus by defaults. Setting it to false will use react-native-modal implemented <View /> and it does not implement trap focus

Besides the fact that this doesn't have the "focus trap", is there any downside to switching to `react-native-modal` from the default `react-native` `Modal` component when we are on web? I am still pretty confused about what this prop even does or why we'd want it:

<img width="934" alt="2022-11-30_14-01-16" src="https://user-images.githubusercontent.com/32969087/204933823-cbd8de76-0030-498a-a282-23d62da5d6f2.png">

**Comment by marcaaron** [comment]
Thanks for the help so far. After reviewing, I don't think the proposal we have here is acceptable yet.

- The performance regression is not _great_. I think we should really try to get to the bottom of that so we can also fix it for the "Search" screen and anything else using the hack in [this comment](https://github.com/Expensify/App/issues/10414#issuecomment-1332905631), but maybe that should be covered in a different issue. 
- It's unclear whether `coverScreen` is needed and in what contexts. I tested out a solution that forces `react-native-modal` with the props `coverScreen: false` and `hasBackdrop: true` and while it fixes the issue the backdrop no longer shows up (maybe I missed something in @s77rt's proposal that addresses this). It's unclear whether disabling the `coverScreen` feature is a safe change or will lead to other issues.

**Comment by s77rt** [comment]
@marcaaron 

- As stated in my proposal, I made the changes to match with the "Search" screen so we can render the input at the first render. If this is causing a lag, this should be handled separably since the lag is also on the "Search" page)
- The backdrop should show up, so you probably forgot something. Changing `coverScreen` to false will actually fix a lot of bugs including this one, and since you are not displaying real modals just context menu it seems logical to not use coverScreen, as for the future bugs, probably accessibility related, you will be able to focus on other elements not on the modal,

Is the change worth it? Safari related issues are always complex and will often lead to hacky solutions, I would say we close the issue (not to be fixed) What concerns me is the modal change, and specially this [bug](https://github.com/Expensify/App/issues/12521#issuecomment-1320362457) the fix of that bug requires same modal modification and if we made those modal modification we may as well continue with the proposed fix above (use autoFocus and shouldShowOptions for didScreenTransitionEnd)

**Comment by parasharrajat** [comment]
The CoverScreen and backdrop prop was added to prevent the context menu from blocking the user interactions with the rest of the app while the context menu is open. This way user can directly open another context menu for a different message without having to first close the previous context menu. 

--- 

Thanks for the details every one. I will check on it. 

Suggested changes for OptionSelector can cause the character typing lag on search input. Another related issue is animation lag for RHN Modal. I will add more details once I have analyzed the details above.

**Comment by marcaaron** [comment]
>Yes, [microtask(queueMicrotask)](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide) can do that for us, it is also underlying layer for react component rendering.
So we can use [focus()](https://github.com/Expensify/App/blob/820e876d5e4a8cbffe272220c1182c8942ed93cf/src/components/OptionsSelector/BaseOptionsSelector.js#L94) to open the keyboard in the componentDidMount, or use autoFocus[1](https://github.com/Expensify/App/issues/10414#user-content-fn-1-1978cf725b9899d37f67522a49d2ac40). Just make sure that they are triggered by the same [execution stack](https://developer.mozilla.org/en-US/docs/Web/JavaScript/EventLoop#stack) of the click handler, not by the end of an animation. (animation generally uses requestAnimationFrame)

Feels like one correct way to do it would be to improve the rendering performance so the async hack is no longer required.

>fakeInput is also a workaround

How does this solve the issue of the "real" input also needing focus? The keyboard would open but a fake input would have focus?

@ntdiary Sorry I'm just not understanding the "focus trap" thing or why it is the cause of the issue.

>`ModalFocusTrap` will always refocus a non input box element

Just to ground this a bit in the current example - which element is getting refocused? And why does it get refocused?

**Comment by marcaaron** [comment]
>Is the change worth it? Safari related issues are always complex and will often lead to hacky solutions, I would say we close the issue (not to be fixed) What concerns me is the modal change, and specially this https://github.com/Expensify/App/issues/12521#issuecomment-1320362457 the fix of that bug requires same modal modification and if we made those modal modification we may as well continue with the proposed fix above (use autoFocus and shouldShowOptions for didScreenTransitionEnd)

@s77rt I don't think closing is an option. We have to do something because we want this to work the same on all platforms. So we must either fix the iOS Safari issue or stop focusing on these screens (but that seems undesirable for the **Change Password** and **New Room** case mentioned in the issue description).

**Comment by marcaaron** [comment]
>The backdrop should show up, so you probably forgot something.

@s77rt can you point me to a branch that I can test? I am testing with a hardware iOS device so that could also be the issue.

**Comment by marcaaron** [comment]
>prevent the context menu from blocking the user interactions with the rest of the app while the context menu is open. This way user can directly open another context menu for a different message without having to first close the previous context menu

@parasharrajat I'm not following. Which user interactions are we trying to allow? And in what case are we opening multiple context menus? Can you use a specific example? Or provide a screenshot so I can better understand why we are using the `coverScreen` prop?

**Comment by s77rt** [comment]
@marcaaron My proposal is old, so you have to use an old version
I just tested with 1.2.27-0 and the backdrop is there.
`git switch --detach tags/1.2.27-0`
`git apply diff.tmp`

**Comment by s77rt** [comment]
@parasharrajat I'm not sure about the coverScreen use case you stated. Is there an issue that we can reproduce if coverScreen is set to false?

**Comment by marcaaron** [comment]
@s77rt Ok I created a draft with the changes applied to the latest version (it needs to work there) want to look and see if I'm missing anything? On my hardware iOS device the backdrop is missing.

![2022-12-01_09-45-56](https://user-images.githubusercontent.com/32969087/205145512-3ef91caf-6943-4250-9685-87e8938e2ae7.png)

https://github.com/Expensify/App/pull/13245

**Comment by s77rt** [comment]
@marcaaron We can fix this later
Let's focus on whether the proposal is good enough or still need more investigation

**Comment by marcaaron** [comment]
I think it needs more investigation.. if we can't implement the fix without causing this regression then we don't have a workable solution. I think this issue is pretty tough though and bot @s77rt and @ntdiary have added crucial context towards solving it and should get some compensation (not really my department - but how I feel about this one so far).

**Comment by marcaaron** [comment]
The missing backdrop for the menu where "New Chat" button lives.

**Comment by s77rt** [comment]
I meant besides that. The backdrop can be fixed easily I think, will look at it.
Any other regression?

**Comment by s77rt** [comment]
### Proposal (Updated)
```diff
diff --git a/src/components/Modal/BaseModal.js b/src/components/Modal/BaseModal.js
index 91b1867a6f..f7300cc098 100644
--- a/src/components/Modal/BaseModal.js
+++ b/src/components/Modal/BaseModal.js
@@ -103,8 +103,8 @@ class BaseModal extends PureComponent {
                 backdropColor={themeColors.modalBackdrop}
                 backdropOpacity={hideBackdrop ? 0 : 0.5}
                 backdropTransitionOutTiming={0}
-                hasBackdrop={this.props.fullscreen}
-                coverScreen={this.props.fullscreen}
+                hasBackdrop={typeof(this.props.hasBackdrop) === "boolean" ? this.props.hasBackdrop : this.props.fullscreen}
+                coverScreen={typeof(this.props.coverScreen) === "boolean" ? this.props.coverScreen : this.props.fullscreen}
                 style={modalStyle}
                 deviceHeight={this.props.windowHeight}
                 deviceWidth={this.props.windowWidth}
diff --git a/src/components/Popover/index.js b/src/components/Popover/index.js
index 9a2e20ed3d..354275282a 100644
--- a/src/components/Popover/index.js
+++ b/src/components/Popover/index.js
@@ -31,6 +31,7 @@ const Popover = (props) => {
             // eslint-disable-next-line react/jsx-props-no-spreading
             {...props}
             fullscreen={props.isSmallScreenWidth ? true : props.fullscreen}
+            hasBackdrop={props.isSmallScreenWidth ? true : props.hasBackdrop}
             animationInTiming={props.disableAnimation && !props.isSmallScreenWidth ? 1 : props.animationInTiming}
             animationOutTiming={props.disableAnimation && !props.isSmallScreenWidth ? 1 : props.animationOutTiming}
         />
diff --git a/src/components/PopoverMenu/BasePopoverMenu.js b/src/components/PopoverMenu/BasePopoverMenu.js
index 04367c66d3..580a5bb6de 100644
--- a/src/components/PopoverMenu/BasePopoverMenu.js
+++ b/src/components/PopoverMenu/BasePopoverMenu.js
@@ -99,6 +99,8 @@ class BasePopoverMenu extends PureComponent {
                 animationInTiming={this.props.animationInTiming}
                 disableAnimation={this.props.disableAnimation}
                 fromSidebarMediumScreen={this.props.fromSidebarMediumScreen}
+                hasBackdrop={this.props.hasBackdrop}
+                coverScreen={this.props.coverScreen}
             >
                 <View style={this.props.isSmallScreenWidth ? {} : styles.createMenuContainer}>
                     {!_.isEmpty(this.props.headerText) && (
diff --git a/src/components/RoomNameInput/index.js b/src/components/RoomNameInput/index.js
index 6c015ec361..33a5ce414c 100644
--- a/src/components/RoomNameInput/index.js
+++ b/src/components/RoomNameInput/index.js
@@ -56,6 +56,7 @@ class RoomNameInput extends Component {
             <TextInput
                 ref={this.props.forwardedRef}
                 disabled={this.props.disabled}
+                autoFocus={this.props.autoFocus}
                 label={this.props.translate('newRoomPage.roomName')}
                 prefixCharacter={CONST.POLICY.ROOM_PREFIX}
                 placeholder={this.props.translate('newRoomPage.social')}
diff --git a/src/components/RoomNameInput/index.native.js b/src/components/RoomNameInput/index.native.js
index 2eb22be46e..d0cc4cdfa6 100644
--- a/src/components/RoomNameInput/index.native.js
+++ b/src/components/RoomNameInput/index.native.js
@@ -27,6 +27,7 @@ class RoomNameInput extends Component {
             <TextInput
                 ref={this.props.forwardedRef}
                 disabled={this.props.disabled}
+                autoFocus={this.props.autoFocus}
                 label={this.props.translate('newRoomPage.roomName')}
                 prefixCharacter={CONST.POLICY.ROOM_PREFIX}
                 placeholder={this.props.translate('newRoomPage.social')}
diff --git a/src/components/RoomNameInput/roomNameInputPropTypes.js b/src/components/RoomNameInput/roomNameInputPropTypes.js
index db5ce4b3c0..0e91d4a4f8 100644
--- a/src/components/RoomNameInput/roomNameInputPropTypes.js
+++ b/src/components/RoomNameInput/roomNameInputPropTypes.js
@@ -11,6 +11,9 @@ const propTypes = {
     /** Whether we should show the input as disabled */
     disabled: PropTypes.bool,
 
+    /** Whether we should auto focus the input */
+    autoFocus: PropTypes.bool,
+
     /** Error text to show */
     errorText: PropTypes.string,
 
@@ -24,6 +27,7 @@ const defaultProps = {
     onChangeText: () => {},
     value: '',
     disabled: false,
+    autoFocus: false,
     errorText: '',
     forwardedRef: () => {},
 };
diff --git a/src/pages/NewChatPage.js b/src/pages/NewChatPage.js
index 5f79c634d9..6f46187d01 100755
--- a/src/pages/NewChatPage.js
+++ b/src/pages/NewChatPage.js
@@ -219,43 +219,41 @@ class NewChatPage extends Component {
                             onCloseButtonPress={() => Navigation.dismissModal(true)}
                         />
                         <View style={[styles.flex1, styles.w100, styles.pRelative]}>
-                            {!didScreenTransitionEnd && <FullScreenLoadingIndicator />}
-                            {didScreenTransitionEnd && (
-                                <OptionsSelector
-                                    canSelectMultipleOptions={this.props.isGroupChat}
-                                    sections={sections}
-                                    selectedOptions={this.state.selectedOptions}
-                                    value={this.state.searchValue}
-                                    onSelectRow={option => (this.props.isGroupChat ? this.toggleOption(option) : this.createChat(option))}
-                                    onChangeText={(searchValue = '') => {
-                                        const {
-                                            recentReports,
-                                            personalDetails,
-                                            userToInvite,
-                                        } = OptionsListUtils.getNewChatOptions(
-                                            this.props.reports,
-                                            this.props.personalDetails,
-                                            this.props.betas,
-                                            searchValue,
-                                            [],
-                                            this.props.isGroupChat ? this.excludedGroupEmails : [],
-                                        );
-                                        this.setState({
-                                            searchValue,
-                                            userToInvite,
-                                            recentReports,
-                                            personalDetails,
-                                        });
-                                    }}
-                                    headerMessage={headerMessage}
-                                    hideAdditionalOptionStates
-                                    forceTextUnreadStyle
-                                    shouldFocusOnSelectRow={this.props.isGroupChat}
-                                    shouldShowConfirmButton={this.props.isGroupChat}
-                                    confirmButtonText={this.props.translate('newChatPage.createGroup')}
-                                    onConfirmSelection={this.props.isGroupChat ? this.createGroup : () => {}}
-                                />
-                            )}
+                            <OptionsSelector
+                                canSelectMultipleOptions={this.props.isGroupChat}
+                                sections={sections}
+                                selectedOptions={this.state.selectedOptions}
+                                value={this.state.searchValue}
+                                onSelectRow={option => (this.props.isGroupChat ? this.toggleOption(option) : this.createChat(option))}
+                                onChangeText={(searchValue = '') => {
+                                    const {
+                                        recentReports,
+                                        personalDetails,
+                                        userToInvite,
+                                    } = OptionsListUtils.getNewChatOptions(
+                                        this.props.reports,
+                                        this.props.personalDetails,
+                                        this.props.betas,
+                                        searchValue,
+                                        [],
+                                        this.props.isGroupChat ? this.excludedGroupEmails : [],
+                                    );
+                                    this.setState({
+                                        searchValue,
+                                        userToInvite,
+                                        recentReports,
+                                        personalDetails,
+                                    });
+                                }}
+                                headerMessage={headerMessage}
+                                hideAdditionalOptionStates
+                                forceTextUnreadStyle
+                                shouldFocusOnSelectRow={this.props.isGroupChat}
+                                shouldShowConfirmButton={this.props.isGroupChat}
+                                confirmButtonText={this.props.translate('newChatPage.createGroup')}
+                                onConfirmSelection={this.props.isGroupChat ? this.createGroup : () => {}}
+                                shouldShowOptions={didScreenTransitionEnd}
+                            />
                         </View>
                     </>
                 )}
diff --git a/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js b/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js
index 465902df23..30b5644382 100644
--- a/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js
+++ b/src/pages/home/sidebar/SidebarScreen/FloatingActionButtonAndPopover.js
@@ -169,6 +169,8 @@ class FloatingActionButtonAndPopover extends React.Component {
                     anchorPosition={styles.createMenuPositionSidebar}
                     onItemSelected={this.hideCreateMenu}
                     fromSidebarMediumScreen={!this.props.isSmallScreenWidth}
+                    hasBackdrop={true}
+                    coverScreen={false}
                     menuItems={[
                         {
                             icon: Expensicons.ChatBubble,
diff --git a/src/pages/settings/PasswordPage.js b/src/pages/settings/PasswordPage.js
index c087bf5931..e1270f29b5 100755
--- a/src/pages/settings/PasswordPage.js
+++ b/src/pages/settings/PasswordPage.js
@@ -181,6 +181,7 @@ class PasswordPage extends Component {
                                         hasError={this.state.errors.currentPassword}
                                         errorText={this.getErrorText('currentPassword')}
                                         onSubmitEditing={this.submit}
+                                        autoFocus
                                     />
                                 </View>
                                 <View style={styles.mb6}>
diff --git a/src/pages/workspace/WorkspaceNewRoomPage.js b/src/pages/workspace/WorkspaceNewRoomPage.js
index 654e2ce8cc..ecc83efd62 100644
--- a/src/pages/workspace/WorkspaceNewRoomPage.js
+++ b/src/pages/workspace/WorkspaceNewRoomPage.js
@@ -163,6 +163,7 @@ class WorkspaceNewRoomPage extends React.Component {
                             errorText={this.state.errors.roomName}
                             onChangeText={roomName => this.clearErrorAndSetValue('roomName', roomName)}
                             value={this.state.roomName}
+                            autoFocus
                         />
                     </View>
                     <View style={styles.mb5}>
```

### Update
Updated to match current latest version

**Comment by marcaaron** [comment]
>This div element with tabindex="0" is getting refocused, because the focusFirstDescendant in trapFocus is trying to focus a focusable descendant.

@ntdiary Good explanation of what happens in the code. Why does it happen? Can we stop it somehow?

**Comment by s77rt** [comment]
@ntdiary The backdrop fix is easy

> Rendering the modal directly in the parent element means that the layout of the descendant elements will always be constrained by the parent element.

That's right, almost,  but we can use CSS position to be set to `fixed` instead of `absolute`

> Absolutely positioned elements are positioned with respect to a containing block, which is the nearest postioned ancestor. If there is no positioned ancestor, the viewport will be the containing block.

> Elements with fixed positioning are fixed with respect to the viewport—the viewport is always their containing block.

https://www.w3.org/wiki/CSS_absolute_and_fixed_positioning?source=post_page#:~:text=Absolutely%20positioned%20elements%20are%20positioned,is%20always%20their%20containing%20block.

Apply this
```diff
diff --git a/node_modules/react-native-modal/dist/modal.style.d.ts b/node_modules/react-native-modal/dist/modal.style.d.ts
index 1734c97..b9d09ad 100644
--- a/node_modules/react-native-modal/dist/modal.style.d.ts
+++ b/node_modules/react-native-modal/dist/modal.style.d.ts
@@ -1,6 +1,6 @@
 declare const _default: {
     backdrop: {
-        position: "absolute";
+        position: "fixed";
         top: number;
         bottom: number;
         left: number;
diff --git a/node_modules/react-native-modal/dist/modal.style.js b/node_modules/react-native-modal/dist/modal.style.js
index 191ea2f..b6aa1ca 100644
--- a/node_modules/react-native-modal/dist/modal.style.js
+++ b/node_modules/react-native-modal/dist/modal.style.js
@@ -1,7 +1,7 @@
 import { StyleSheet } from 'react-native';
 export default StyleSheet.create({
     backdrop: {
-        position: 'absolute',
+        position: 'fixed',
         top: 0,
         bottom: 0,
         left: 0,
```

**Comment by sketchydroide** [comment]
Thanks for jumping in @marcaaron, this seems like a particualrly complex problem, and some of this things go over my head a bit, I can understand the overall concept but the details are hard to follow

**Comment by s77rt** [comment]
@sketchydroide 
1. Make changes to the modal to avoid focus trap because focus trap make the input un-focus-able.
2. Make changes to render and focus the textinput in the first render call to make it a part of the call-stack that originated from the user interaction because Safari only allow focus calls from call-stacks that originated from the user interaction (such as a click event)

**Comment by sketchydroide** [comment]
@s77rt thanks that makes it very clear

**Comment by s77rt** [comment]
hasBackdrop + css position fixed = hasBackdrop that covers the screen.


That's what we want, having the backdrop cover the screen but without using RN's modal.

___

I think the issue is clear for everyone now and so the fix.

**Comment by s77rt** [comment]
@ntdiary Looking good
can you test against this [bug](https://github.com/Expensify/App/issues/12521#issuecomment-1320362457) too?

**Comment by sketchydroide** [comment]
@s77rt and @ntdiary can one of you (or both) explain what it the purpose of the `focus trap`
I come from iOS development and this is not a concept there, to better understand what work best for us it would be nice to have a better explanation of what is exactly involved in the solution

**Comment by s77rt** [comment]
@sketchydroide focus trap is introduced for accessibility.
You shouldn't be able to focus on elements that are behind a modal.

This is causing a problem so we are looking to disable it.

Thats because we are trying to focus on the input while the modal is still open.
You may ask then why we are focusing in the input while the modal is still open. Well we can't wait for the modal to close because doing so means using another call-stack. Another solution is to disable animation.

But disabling animation is not enough with the current code because we are rendering the text input then closing the modal.

If we want to solve this issue by disabling animation we will have to close the modal then render the input. This way we won't have to deal with focus trap.

**Comment by marcaaron** [comment]
Sounds like there are 2 main solutions

## Solution 1

### Use `react-native-modal` instead of the `react-native-web` `Modal` to avoid the "focus trap"

This solution comes with a problem in that the backdrop moves off screen and will need some kind of style change to fix. Either in `react-native-modal` or our own code?

>actually after [this commit](https://github.com/Expensify/App/commit/f07542c634d38151eb2bb5ced327cacdd8054d5b), we move the FloatingActionButton component into the FloatingActionButtonAndPopover component. This has changed the parent element of the popover menu modal, and new parent element is positioned at the bottom of the screen and has a height of zero. This makes the backdrop offscreen.

So do we just need to revert something about that change? Add a style in our own App? Or must the change happen in `react-native-modal`? 

## Solution 2
### Continue to use `react-native-web` `Modal`, but fix the "focus trap" issue

This would require opening a bug report and suggest a change to `react-native-web` allow us to disable the "focus trap" behavior.

## Remaining Questions / Next Steps

What is a focus trap? Why does it exist?

I think this is a good place to start because the best change would be to prevent the behavior in `react-native-web` if we can understand the intention of it and whether it's something we should change or not.

Once we answer that then we can figure out whether:

- [ ] Find out if this something we can/should change or not
- [ ] If not, explore using `react-native-modal` instead and fix the styles in the app or propose some change to `react-native-modal`

**Comment by s77rt** [comment]
@marcaaron 
> So do we just need to revert something about that change?

No, this won't help. The style still must be applied..

> What is a focus trap? Why does it exist?

https://github.com/Expensify/App/issues/10414#issuecomment-1335532039


Maybe we have a third solution by disabling animation and change the order of events.


I'm leaning toward solution one because it will fix other issues.

Both solution one and two will lead to an accessibility problem, which is: you will be able to focus on elements that are behind the modal, and that's the issue that trap focus fix. So it's either one issue or another.

Solution three won't deal with trap focus, so maybe we should look at this too

___

Solution 1: Fix this issue and other issues. Introduce an accessibility problem
Solution 2: Fix this issue. Introduce an accessibility problem
Solution 3: Fix this issue. Disable animation (at least the closing modal animation)

**Comment by sketchydroide** [comment]
@s77rt thanks for the explanation on the `focus trap` this makes more sense.
Wich animation do you mean, the modal one, I don't think we wan't to disable that one.

**Comment by sketchydroide** [comment]
@ntdiary I think we also want to enable focus trap after the event, so that the acessibility still works.
Also this seem to be less of a problem in mobile, I mean I know people still use accessibility in mobile, but not as much as they do in Web or Desktop. In particular the example with the tabs you gave.

As for the Portal use I think that is a way broader change that although we can address it won't be to fix this bug, we can try to porpose that change in the #expensify_open_source slack channel. Not sure what traction it will have

**Comment by sketchydroide** [comment]
my concern is that focus trap does not become enabled, again, if it's just during the animation it should be ok.

**Comment by s77rt** [comment]
@sketchydroide Yes I'm referring to the modal animations. We can disable the modal close animation only and only on Safari. But this won't be straightforward as we have to make PR for react-native-modal because there is no option to disable animation.

I think I may have a solution that is more satisfying, will get back to this soon

**Comment by sketchydroide** [comment]
We would prefer to keep all the animations, and mainly keep the behaviour consistent across all platforms if possible.
Also if possible to have a solution whitouth rellying on a PR getting approved outside of the Expensify repos would be preferable, as we can make sure the PRs are taken care faster.

**Comment by s77rt** [comment]
@marcaaron
There are two root issues:
1. trap focus which is on RN, either fix RN modal or replace it,
2. render and focus must be done on the same call-stack that originated from user interaction, this must be fixed in the app, pretty straightforward fix as stated above


Here a regression example, on another modal but same thing, pressing TAB will focus on other elements beside the one on the modal. On mobile this is less of an issue, so there is a probability that we could just disable trap focus on mobile and move on.
https://github.com/Expensify/App/issues/12521#issuecomment-1318643369

___

Let's only focus on the first root cause as the second is not that hard to figure out.
Here are what we can do:
1. Disable trap focus on mobile by using `react-native-modal` modal
2. Keep using RN modal but make a change to disable trap focus
4. Keep using RN modal, make a change to make it work but keep trap focus working

IMO the best solution is solution 1. and here is why:
1. Fixes this issue
2. The regression is unnoticeable (mobile users does not have "TAB accessibility")
3. Fix this annoying bug https://github.com/Expensify/App/issues/12521#issuecomment-1320362457 It's related to navigation and such issues are always quite hard to fix, having it's own separate issue won't get us far I think, and that's why I'm listing it here as bonus point for solution 1

**Comment by marcaaron** [comment]
It feels like we're going around in circles a bit. We can keep talking, sharing opinions on what the "correct" fix, and repeat what has been said already, but it's not going to get us any closer to a solution. @sketchydroide I think bringing a summarized conversation to Slack is the best way to get forward momentum here.

**Comment by s77rt** [comment]
@marcaaron I have already provided a solution a while ago and updated it so it can be tested with the latest app version.
I'm actually waiting for your review...

I have provided alternative solutions and tried to explain the root issue just to get everyone on track to help better evaluate my proposal

**Comment by marcaaron** [comment]
Thanks for the thoughts. Gonna pause the discussion here for a moment so we can continue in Slack.

**Comment by sketchydroide** [comment]
currently still being discussed on slack, but seems we at least have a clear direction of what to do next

**Comment by JmillsExpensify** [comment]
[Slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1670271723022949) is here by the way.

**Comment by JmillsExpensify** [comment]
Quickly summarizing a key point from the linked thread, we're going to wait to hear back from the upstream maintainer. Issue is on hold as a result.

**Comment by sketchydroide** [comment]
what @JmillsExpensify said, not overdue, we are waiting
[Here is a link to the upstream mantainer](https://github.com/necolas/react-native-web/issues/2175#issuecomment-1339212802) request that was made

**Comment by trjExpensify** [comment]
Yep, makes sense to make this a `weekly`. Updated.

**Comment by sketchydroide** [comment]
I think this is still waiting?

**Comment by JmillsExpensify** [comment]
Yes, still working on an upstream fix. The maintainer has noted that he has a long pipeline of improvements that he's working on.

**Comment by sketchydroide** [comment]
It think this is still the same

**Comment by trjExpensify** [comment]
Are we not exploring our existing fork in the meantime then, as per the updates we made to [the process](https://stackoverflowteams.com/c/expensify/questions/12761)?

**Comment by sketchydroide** [comment]
not sure, I think it's definetly a possibility, @marcaaron what do you think?

**Comment by JmillsExpensify** [comment]
Great, let's bring the discussion to Slack. @sketchydroide are you heading that up? For what it's worth, if we are following our own process at this point, the process says fork. But I think a discussion would be good regardless, let's do it.

**Comment by sketchydroide** [comment]
just cameback from OOO yesterday, I'll create the thread in slack

**Comment by sketchydroide** [comment]
Slack thread created [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1674129723623539)

**Comment by sketchydroide** [comment]
I think the general idea from the conversation in slack, is to continue to pursue the change in the upstream while we implement it on our fork

**Comment by trjExpensify** [comment]
Cool, so what are the next steps on getting a PR up for a patch in our RNW fork?

**Comment by kevinksullivan** [comment]
@sketchydroide checking in on this

**Comment by sketchydroide** [comment]
will get this done tomorrow, also maybe we can make this a daily again?

**Comment by kevinksullivan** [comment]
@parasharrajat Andre is no longer working at Expensify so we will want to revisit internal assignment. I re-read the [latest thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1674129723623539) on this and I'm still not totally sure what next steps would be. Do you have an understanding of what our approach is?

CC @marcaaron since you were weighing in on the discussion in slack as well.

**Comment by marcaaron** [comment]
Uh yea so IIRC there were two options for fixing this. Both require we do a patch or fork to either `react-native-modal` or `react-native-web`. Both of them are kind of "hacks" (maintainers not explicitly agreeing with proposed changes). 

The options are:

1. Pick one of the hacks and implement it in a fork (the `react-native-modal` one makes slightly more sense to me)
2. Don't pick any hacks and do a more holistic solution by implementing our own modal on web (we would be able to define it's behavior completely).

At a high level, there is not much value in doing 2 outside of this weird iOS Safari "focus trap" issue. It's a rare case where I'd lean towards maintaining the status quo and picking the "hack" (which IMO isn't a hack at all from _our_ perspective). There is little value in ripping this up and starting over from scratch and there is always a chance that we get the change merged upstream eventually.

Basically, I propose we hire @ntdiary since they have been active in pursuing the solution and promoting the changes to the maintainer (even if they haven't been responsive) and I think the solution is easy enough to understand.

Sorry, I haven't really had a chance to go back and look at all the context here or acknowledge everyone, but @s77rt was very active in trying to help us find a solution so I would propose we split up the C+ duties and have everyone tackle this together and get everyone some bread for helping push it over the finish line.

**Comment by marcaaron** [comment]
Unless we decided to fix it elsewhere actually? Why is there a hold on the issue? 😄

**Comment by trjExpensify** [comment]
>Unless we decided to fix it elsewhere actually? Why is there a hold on the issue? 😄

I don't think we did. That's pointing to [this RNW issue](https://github.com/necolas/react-native-web/issues/2175#issuecomment-1339212802) upstream when we were hopeful of some traction.. months ago.

**Comment by marcaaron** [comment]
Ah, ok cool, then my comment stands as there was no movement there. Let's try to hit up the maintainer.

**Comment by kevinksullivan** [comment]
Just looking back at Marc's [comment](https://github.com/Expensify/App/issues/10414#issuecomment-1457463813). Is our plan to send an offer to @ntdiary and add @s77rt for C+?

**Comment by trjExpensify** [comment]
@parasharrajat, I just tested and it's not fixed. Can you elaborate on how that PR relates to this issue? 

@ntdiary can you bring everyone up-to-speed with where you're at with finding a solution to this problem? I see you've had some dialogue in RNW and the issue is closed. Are we now going the route of a patch to our RNW fork, or something else?

**Comment by marcaaron** [comment]
IMO we should just pass it to an expert contributor to look into as a longer term project rather than assign `Internal`. Probably the only acceptable non-hack solution is to implement a version of `Modal` on web that doesn't have a focus trap or whatever. I don't feel like we are bound to using the React Native Web version of `Modal` or any other modal library. But making a hack to a fork that will never get merged upstream sounds like a bad idea.

**Comment by marcaaron** [comment]
I think we will have better luck building our own `Modal` for mobile web use to workaround the Safari inconsistencies. But open to other opinions. We are not beholden to using any particular modal component and should be able to design one that allows for autofocusing an input (at least I can't think of why we wouldn't be able to).

**Comment by s77rt** [comment]
### Regarding the trap focus:
What if we move the focus to be after the modal is closed?

**Current:**
1. Open FAB
2. Click New Chat
3. New Chat page is opened
4. Input is focused (but reverted due to Trap Focus)
5. Modal is closed.

**Suggested:**
1. Open FAB
2. Click New Chat
3. Close Modal
4. New Chat page is opened
5. Input is focused

I feel like I suggested this before at some point but not sure :sweat_smile:.


### Regarding focus being in the same callstack (after user interaction)
There is no way out of this, we can just make the necessary changes i.e. TextInput must be rendered at the initial render and have autoFocus prop.

**Comment by s77rt** [comment]
Hey, just found that we are actually doing that ^ This PR https://github.com/Expensify/App/pull/12031 made those changes, so now we close the modal first then open chat so Focus Trap is technically avoided :tada:. ...Well, almost, there is another bug on [react-native-modal](https://github.com/react-native-modal/react-native-modal/pull/733) which is that the `onModalHide` callback is executed a little early, we are working on that here https://github.com/Expensify/App/issues/14848.

**Comment by s77rt** [comment]
## Recape
On Safari Keyboard does not show up for two reasons:
1. Calling `focus()` on an element outside an open modal will be reverted due to that modal focus trap implementation.
2. Calling `focus()` must be done in the same callstack of the click event. This is enforced by Safari, focus only works for events that are originating from user interaction.

### Where are we at?
1. Is basically fixed (for all browsers) by [#12031](https://github.com/Expensify/App/pull/12031), now we close modal first before doing anything else.
2. This became harder / impossible since the page is opened from the `onModalHide` callback which is not considered as user interaction.

### Next steps
Unless we are willing to revert https://github.com/Expensify/App/pull/12031 and/or implement our own focus-trap-free modal then I think we should do nothing.

I don't think `VirtualKeyboard` is a good solution here since it will force us to manually control the keyboard state for all inputs i.e. focus will not be enough, we will have to both focus and show keyboard independently. Also I doubt Safari will implement `VirtualKeyboard.show` method since it contradicts with their design logic. If they ever do that then the focus limitation may be removed as well.

**Comment by MelvinBot** [comment]
@kevinksullivan, @parasharrajat Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by trjExpensify** [comment]
>Is basically fixed (for all browsers) by https://github.com/Expensify/App/pull/12031,

but not Safari, correct? 

>Unless we are willing to revert https://github.com/Expensify/App/pull/12031 and/or implement our own focus-trap-free modal then I think we should do nothing.

I like the sound of exploring the latter in more depth. Similar to @marcaaron's thought here: 

>I think we will have better luck building our own Modal for mobile web use to workaround the Safari inconsistencies.

We can't be the only ones that run into this problem, and these are some pretty central flows in the App for requesting money, starting a new chat etc.

**Comment by s77rt** [comment]
@tjferriss 
> but not Safari, correct?

Including Safari. Now focus trap is no longer a problem (well almost as explained here https://github.com/Expensify/App/issues/10414#issuecomment-1509894483).

___

The problem now is about focus not being called after user interaction, since this is a browser limitation we may never holistically fix that. We will always get recurring issues in the future if we call focus in any other callback than the click callback. I'm also curious how others handle this :thinking:.

**Comment by tjferriss** [comment]
^ @trjExpensify

**Comment by s77rt** [comment]
@tjferriss Sorry about that :sweat_smile:  and thanks for the correct tag.

**Comment by MelvinBot** [comment]
@kevinksullivan, @parasharrajat Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@kevinksullivan, @parasharrajat 6 days overdue. This is scarier than being forced to listen to Vogon poetry!

**Comment by trjExpensify** [comment]
>I'm also curious how others handle this 🤔.

Same!

**Comment by jayeshmangwani** [comment]
> @parasharrajat really sorry. That was for notifications and was just a bad way to do it, will not happen again

@huzaifa-99 Rather than commenting on issues to get the notification, Instead, you can subscribe to any Github Issue
<img width="357" alt="Screenshot 2023-04-30 at 4 10 56 AM" src="https://user-images.githubusercontent.com/35371050/235327000-88071593-4935-4cbe-8df7-7df4a9ce881d.png">

**Comment by MelvinBot** [comment]
@kevinksullivan, @parasharrajat Huh... This is 4 days overdue. Who can take care of this?

**Comment by MelvinBot** [comment]
@kevinksullivan, @parasharrajat Huh... This is 4 days overdue. Who can take care of this?

**Comment by kevinksullivan** [comment]
Discussing next steps

https://expensify.slack.com/archives/C01SKUP7QR0/p1683149344735549

**Comment by kevinksullivan** [comment]
Likely going to have callstack take this one, and discussing it here

https://expensify.slack.com/archives/C03UK30EA1Z/p1683152778085989

**Comment by kevinksullivan** [comment]
Thanks for digging into that @narefyev91 ! Going to discuss this a bit internally to determine how we want to move forward

https://expensify.slack.com/archives/C01SKUP7QR0/p1683937815516789?thread_ts=1683149344.735549&cid=C01SKUP7QR0

**Comment by narefyev91** [comment]
> 

The issue is about 2 cases - initial focus and come back focus.
For initial focus it may be resolved by writing own WebView injection - which may work/not work (based on some discussion from here https://stackoverflow.com/questions/32449870/programmatically-focus-on-a-form-in-a-webview-wkwebview) applying those changes inside Native WebView maybe blocked by apple but not really sure.
For second case - when you try to focus input on closing modal - this is possible based on this https://github.com/react-native-modal/react-native-modal/pull/733 - but seems no one will merge that - and possible solution is to fork and apply those fixes which @s77rt suggested. But it will work only when you have interaction with modal. On any other screens like sending money - when you change currency and go back - you will not be able to focus an input again - because it will be 0 user interaction.

**Comment by kevinksullivan** [comment]
discussing 

https://expensify.slack.com/archives/C01SKUP7QR0/p1684373167911069?thread_ts=1683149344.735549&cid=C01SKUP7QR0

**Comment by kevinksullivan** [comment]
still following up here

https://expensify.slack.com/archives/C049HHMV9SM/p1684373273406319?thread_ts=1683649359.591549&cid=C049HHMV9SM

Unassigning as I will be OOO for the next few days.

**Comment by marcaaron** [comment]
>The issue is about 2 cases - initial focus and come back focus

I'm not sure we care about "come back focus"? Is it a requirement here? How do we know you want to interact with an input if you're in the process of "going back"? Maybe you're trying to get to another screen and auto-focus is unnecessary.

Let's just focus on the one case described in the issue description for now (and worry about this "go back" case later)

>For initial focus it may be resolved by writing own WebView injection - which may work/not work (based on some discussion from here https://stackoverflow.com/questions/32449870/programmatically-focus-on-a-form-in-a-webview-wkwebview) applying those changes inside Native WebView maybe blocked by apple but not really sure.

I'm not following this suggestion. To clarify, again, the issue is affecting Safari web browsers on iOS devices not native platforms. AFAICT auto-focus works fine on native devices (and every other platform we've tested).

**Comment by marcaaron** [comment]
@parasharrajat are you still helping to manage this issue? curious to get your thoughts in this conversation. 🙏

**Comment by narefyev91** [comment]
> >The issue is about 2 cases - initial focus and come back focus
> 
> 
> 
> I'm not sure we care about "come back focus"? Is it a requirement here? How do we know you want to interact with an input if you're in the process of "going back"? Maybe you're trying to get to another screen and auto-focus is unnecessary.
> 
> 
> 
> Let's just focus on the one case described in the issue description for now (and worry about this "go back" case later)
> 
> 
> 
> >For initial focus it may be resolved by writing own WebView injection - which may work/not work (based on some discussion from here https://stackoverflow.com/questions/32449870/programmatically-focus-on-a-form-in-a-webview-wkwebview) applying those changes inside Native WebView maybe blocked by apple but not really sure.
> 
> 
> 
> I'm not following this suggestion. To clarify, again, the issue is affecting Safari web browsers on iOS devices not native platforms. AFAICT auto-focus works fine on native devices (and every other platform we've tested).

Yup the main issue was initial focus on safari web, but during discussion one more issue raised - go back focus, and as I remember such issue was raised in some other GitHub issues. I only summarised what currently happened in the discussion here. Regarding some information for injection - to make auto-focus working - I only wanted to indicate that even using WebView and have ability to control render process inside native code (both webview and safari use webkit engine)- people was not able to make autofocus working. For sure native apps and all other platforms do not have current issue which we still discussing here. Hope that helps @marcaaron @puneetlath

**Comment by kevinksullivan** [comment]
Hi @parasharrajat any update on this?

**Comment by kevinksullivan** [comment]
Also discussing picking up SWM to investigate this further here

https://expensify.slack.com/archives/C04878MDF34/p1685549784071119

**Comment by jczekalski** [comment]
Still not sure what to do here. @parasharrajat Are you also looking into this at the moment? Do you have any thoughts on how to resolve this problem?

**Comment by kevinksullivan** [comment]
checking on the status with @jczekalski 

https://expensify.slack.com/archives/C04878MDF34/p1686758234405789?thread_ts=1685549784.071119&cid=C04878MDF34

**Comment by kevinksullivan** [comment]
@jczekalski does Mariusz have a GH handle so I can assign them the issue?

**Comment by jczekalski** [comment]
> @jczekalski does Mariusz have a GH handle so I can assign them the issue?

@kevinksullivan I just DMed him to comment in the issue.

**Comment by staszekscp** [comment]
Hey @kevinksullivan! I'm looking at this issue right now, so you may assign me :)

Also, I'd like to mention, that I noticed a bug that had already been described [here](https://github.com/Expensify/App/issues/21363). However, it also mentions that another, similar task was closed, because the behaviour I am actually trying to fix [was not considered as a bug](https://github.com/Expensify/App/issues/19290#issuecomment-1557625675).

That's why I'd like to clarify if this task is still considered as an issue - should the input autofocus automatically, or should we let users focus it whenever they prefer to do it? In any case there would be some work to be done in order to align the behaviour on all the platforms.

**Comment by kevinksullivan** [comment]
Thanks @staszekscp !

>Also, I'd like to mention, that I noticed a bug that had already been described https://github.com/Expensify/App/issues/21363. However, it also mentions that another, similar task was closed, because the behaviour I am actually trying to fix https://github.com/Expensify/App/issues/19290#issuecomment-1557625675.

Since then we've agreed this a bug that is worth fixing.

>should the input autofocus automatically, or should we let users focus it whenever they prefer to do it? In any case there would be some work to be done in order to align the behaviour on all the platforms.

the input should focus automatically

**Comment by staszekscp** [comment]
Thank you for the answer @kevinksullivan!

**Comment by staszekscp** [comment]
# Proposal

After some investigation I came up with solution that would fix the `mSafari` issue with `autoFocus` while navigating from the `PopoverMenu`.

It basically consists of two simple changes:
1. Move logic from `onModalHide` to `useEffect` in order to combine so called user interaction with input focus
2. Disabling `coverScreen` prop in order to create Modal with a `View` component, and defining a simple `customBackdrop`

You can have a look at it [here](https://github.com/staszekscp/App/pull/27).

I was also investigating focusing the input going back to the screen, but unfortunately I haven't come up with anything yet. 😢 

Let me know what you think 😄

**Comment by kevinksullivan** [comment]
@parasharrajat since this is one of the longest running bugs, I'd appreciate you prioritizing the review! Let me know when you can do so.

**Comment by kevinksullivan** [comment]
Hi @parasharrajat mind providing an update on this?

**Comment by staszekscp** [comment]
Hello, do you have any thoughts about my proposal? It's been 3 weeks already since it had been posted, and no feedback/review was provided 😄

**Comment by ArekChr** [comment]
Hey there, I will help with reviewing proposals as C+. Please assign me

**Comment by ArekChr** [comment]
Reviewing and testing https://github.com/Expensify/App/issues/10414#issuecomment-1621710184

**Comment by ArekChr** [comment]
@staszekscp, I've tested your solution, it works great on mobile Safari there are no issues with the keyboard, but I've also found a regression on the web and desktop versions. After navigating from a new group or new chat, the input is no longer focused, could you include a fix of that regression?
Please also update your proposal using the provided [template](https://github.com/Expensify/App/blob/main/contributingGuides/PROPOSAL_TEMPLATE.md). This will ensure we are on the same page. The issues might have changed slightly because there is no longer a 'new room' in the bottom sheet. Overall, however, I'm pretty pleased with the solution.

**Comment by staszekscp** [comment]
@ArekChr Thanks for the feedback! I've just fixed the problem that you've mentioned - it seems to be working correctly on all platforms, but feel free to check this out :) I've also updated the proposal using the provided template. Thanks again!

**Comment by staszekscp** [comment]
Hey @ArekChr! I've just posted the [PR](https://github.com/Expensify/App/pull/23922) - I had to fix a small regression today, but now I think it works as expected :) Could you do the review?

**Comment by ArekChr** [comment]
Thanks for the update, @staszekscp. Overall the proposal looks good to me now! https://github.com/Expensify/App/issues/10414#issuecomment-1621710184
:ribbon: :eyes: :ribbon: C+ reviewed

**Comment by kevinksullivan** [comment]
What's the latest on this @staszekscp ?

**Comment by staszekscp** [comment]
Hey @kevinksullivan! I’ve applied changes yesterday in the morning, and I’m waiting for feedback :)

**Comment by kevinksullivan** [comment]
Got it, thanks for the update! Are we waiting on @parasharrajat then?

**Comment by sakluger** [comment]
@parasharrajat @aldo-expensify do you know if this PR will fix https://github.com/Expensify/App/issues/25477?

**Comment by sakluger** [comment]
Bump @parasharrajat @aldo-expensify, I put https://github.com/Expensify/App/issues/25477 on hold in case this one fixes it. Do you think this PR will fix that issue?

**Comment by sakluger** [comment]
Just checking in. It looks like:

- The [PR](https://github.com/Expensify/App/pull/23922) was closed because an edge case was discovered that caused the fix to not work
- @staszekscp is investigating other fixes, but their top focus is another Wave 10 issue, so this issue is waiting until they can re-prioritize this one.

Is that the latest? Thanks!

**Comment by aldo-expensify** [comment]
That is my understanding too

**Comment by sakluger** [comment]
Any updates on this one, or is it still deprioritized?

**Comment by sakluger** [comment]
@kevinksullivan do you still think we should fix this one? And is it still an $8k bug? Just wanted to check in since it's been deprioritized since around October now.

**Comment by kevinksullivan** [comment]
Changing price in GH and upwork this has not been a priority. @kbecciv can you please confirm you can reproduce this still? Then we can source volunteers if necessary.

**Comment by eVoloshchak** [comment]
Can confirm this is still reproducible on mWeb Safari for some inputs


https://github.com/Expensify/App/assets/9059945/9c245eba-fdb3-4cb2-af6b-81d3da8eb7b3

**Comment by sakluger** [comment]
@kevinksullivan fyi I added https://github.com/Expensify/App/issues/25477 (which is held on this issue) to the vip-vsb project and I think this should be added to the same project since it's chat-related functionality.

**Comment by sakluger** [comment]
Any updates here?

**Comment by sakluger** [comment]
Friendlly bump - any updates?

**Comment by sakluger** [comment]
@parasharrajat are you suggesting that we discuss whether we want to still solve this or just close the issue?

**Comment by kevinksullivan** [comment]
@sakluger I will loop you in as BZ here if you don't mind, as I'm going OOO for a bit.

**Comment by sakluger** [comment]
Sounds good. I'll ask @parasharrajat in Slack if we should close this issue.

**Comment by sakluger** [comment]
We're going to close out the issue. Given all the work that's been done, @parasharrajat is eligible for the full $250 payment. I'll post a payment summary below.

**Comment by sakluger** [comment]
Summarizing payment on this issue:

Contributor+: @parasharrajat $250, please request on Newdot

**Comment by JmillsExpensify** [comment]
$250 approved for @parasharrajat
